### PR TITLE
FF-A: spmc_sp_handler fixes

### PR DESCRIPTION
**Pull request by jellesels-arm**
Found some small errors in the FF-A spmc_sp_handler.
This Pull request fixes these.

### CODE REVIEW

File: core/arch/arm/kernel/thread.c [hunk 1]
```diff
@@ -1495,7 +1495,7 @@ uint32_t thread_enter_user_mode(unsigned long a0, unsigned long a1,
 	 */
 	regs = thread_get_ctx_regs();
 	set_ctx_regs(regs, a0, a1, a2, a3, user_sp, entry_func, spsr);
-	rc = __thread_enter_user_mode(regs, exit_status0, exit_status1);
+	rc = thread_enter_user_mode_ctx(regs, exit_status0, exit_status1);
```
[the remark this answers is not in the document]
Comment (jellesels-arm) [reply core/arch/arm/kernel/thread.c:1498]: __thread_enter_user_mode was defined before inside thread_private.h. secure_partition.c doesn't have access to thread_private.h anymore as it was move outside the core/arch/arm/kernel. I renamed it to make it clear that it's not a 'private' function anymore. thread_enter_user_mode was already used as a function

File: core/arch/arm/kernel/spmc_sp_handler.c [hunk 2]
```diff
@@ -267,13 +267,21 @@ void spmc_sp_msg_handler(struct thread_smc_args *args,
...
 		case FFA_FEATURES:
 			handle_features(args);
 			sp_enter(args, caller_sp);
 			break;
+		case FFA_PARTITION_INFO_GET:
+			ts_push_current_session(&caller_sp->ts_sess);
+			spmc_handle_partition_info_get(args, &caller_sp->rxtx);
+			ts_pop_current_session();
+			sp_enter(args, caller_sp);
+			break;
```
[the remark this answers is not in the document]
Comment (jellesels-arm) [reply core/arch/arm/kernel/spmc_sp_handler.c:284]: This snuck into it somehow. Fixed it

**Comment by jellesels-arm** [comment]
> The bugs in `spmc_sp_msg_handler()` should have been obvious with some testing. I think the FF-A stuff is becoming complex enough that we need regression testing in the CI loop.

We are working on some tests that we can add to xtest. At the moment we are still figuring out how we can add the necessary SPs to be included in the build. SPs are build as part of the trusted-services project.

**Comment by jforissier** [comment]
> > The bugs in `spmc_sp_msg_handler()` should have been obvious with some testing. I think the FF-A stuff is becoming complex enough that we need regression testing in the CI loop.
> 
> We are working on some tests that we can add to xtest. At the moment we are still figuring out how we can add the necessary SPs to be included in the build. SPs are build as part of the trusted-services project.

Awesome! Perhaps the trusted-services project could simply be added to our manifest.git/build.git?

**Comment by jforissier** [comment]
- For commit "core: kernel : move secure_partition.c"
I am trying to understand the justification for this commit. You mention some dependency in the commit description but the real reason should be that it is not architecture-specific. Why not re-use the same message template as the commit you are citing? Also note that the preferred way to cite a previous commit in the same project is: `commit <sha1-12-char> "Commit subject"`. For this I use an alias in my ~/.gitconfig: `cite = show --abbrev=12 --format='commit %h (\"%s\")' -s`.

- For commit "FF-A: Fix spmc_sp_msg_handler bugs"
I would prefer the following subject and description:
  ```
  core: FF-A: add missing break statements to spmc_sp_msg_handler()

  Some break statements are missing in spmc_sp_msg_handler(). Add them.
  ```

With that,
`Acked-by: Jerome Forissier <jerome@forissier.org>`

**Comment by jenswi-linaro** [comment]
I'm not sure that secure_partition.c should be moved to core/kernel. Isn't SPs per definition a very Arm specific thing?

**Comment by jforissier** [comment]
> I'm not sure that secure_partition.c should be moved to core/kernel. Isn't SPs per definition a very Arm specific thing?

My knowledge of FF-A is probably not good enough to say for sure, but FF-A seems to be mostly a protocol, isn't it? The underlying (arch-specific) isolation mechanisms are pretty much irrelevant, aren't they?

File: core/arch/arm/kernel/thread.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jellesels-arm) [reply core/arch/arm/kernel/thread.c]: Dropped the patch

**Comment by jellesels-arm** [comment]
> > I'm not sure that secure_partition.c should be moved to core/kernel. Isn't SPs per definition a very Arm specific thing?
> 
> My knowledge of FF-A is probably not good enough to say for sure, but FF-A seems to be mostly a protocol, isn't it? The underlying (arch-specific) isolation mechanisms are pretty much irrelevant, aren't they?
> 
> **Edit**: on the other hand I would also be OK with the alternative approach, i.e., wait until another architecture decides to use this code before moving it out of `arch/arm`.

For the moment I dropped the commit. We can move it once it supported by a different architecture.

**Comment by jforissier** [comment]
@jellesels-arm fine with me. Then I just need to repeat my previous comments ;-)

>     * For commit "FF-A: Fix spmc_sp_msg_handler bugs"
>       I would prefer the following subject and description:
>       ```
>       core: FF-A: add missing break statements to spmc_sp_msg_handler()
>       
>       Some break statements are missing in spmc_sp_msg_handler(). Add them.
>       ```
> 
> 
> With that,
> `Acked-by: Jerome Forissier <jerome@forissier.org>`

**Pull request merged**